iso15118/secc/states/iso15118_2_states.py

```diff
@@ -1458,6 +1458,33 @@ async def process_message(
         #     )
         #     return
 
+        # [V2G2-225] The SECC shall send the negative ResponseCode
+        # FAILED_ChargingProfileInvalid in
+        # the PowerDelivery response message if the EVCC sends a ChargingProfile which
+        # is not adhering to the PMax values of all PMaxScheduleEntry elements according
+        # to the chosen SAScheduleTuple element in the last ChargeParameterDiscoveryRes
+        # message sent by the SECC.
+        if power_delivery_req.charging_profile:
+            for profile_entry in power_delivery_req.charging_profile.profile_entries:
+                for schedule in self.comm_session.offered_schedules:
+                    for p_max_profile_entry in schedule.p_max_schedule.schedule_entries:
+                        if (
+                            p_max_profile_entry.time_interval.start
+                            == profile_entry.start
+                        ):
```

Comment:
I think the check should be made against the chosen schedule in the offered schedule.
The following case is an example use case where the above check could pass but should fail:
Offered chosen schedule
(0s)<---5KW--->(10s)<---10KW<---(60s)--->End

Returned profile
(0s)<--4KW-->(2s)<--6KW-->(3s)<--10KW-->(60s)--->End

So it is necessary to pick every interval returned in the charging profile and make sure it stays below the limits of the interval(s) that entry might correspond to.

Reply:
I think the solution is something like this
```python
for schedule in self.comm_session.offered_schedules: 
    if schedule.sa_schedule_tuple_id == power_delivery_req.sa_schedule_tuple_id:
        sa_schedule_pmax = max([profile_entry.p_max.value * 10**profile_entry.p_max.multiplier for profile_entry in schedule.p_max_schedule.schedule_entries])
        ev_profile_pmax = max([profile_entry.max_power.value * 10**profile_entry.max_power.multiplier for profile_entry in power_delivery_req.charging_profile.profile_entries])
        if ev_profile_pmax > sa_schedule_pmax:
            self.stop_state_machine(....)
            return
```

I guess the code to find the max of each set could be optimized, maybe using `map` or something,,,

Reply:
@tropxy should not we check each profile entry values according to time, and it should be lower than what offered in offered_schedules. if we only check the maximum value of given schedule, it would be invalid for the case below:

Offered chosen schedule
(0s)<---5KW--->(10s)<---10KW<---(60s)--->End

Returned profile
(0s)<--4KW-->(2s)<--6KW-->(3s)<--10KW-->(60s)--->End 

in this case, for the value 6KW should not exceed 5KW. This is what I understood.


do you think if this is a realistic case?

Reply:
I have added a logic to cover the case above, but still, I am not happy with it...

Reply:
yeah, looking at this again, you have a point. I believe you are right, we have to go over the different slots of time of the chosen schedule (sa_schedule) and see if the power consumed by the EV (ev_profile) during that slot does not exceed the one set by the sa_schedule.

Thus, the logic I provided before is incorrect...

Reply:
This shall be the logic then, I believe....
```python
for schedule in self.comm_session.offered_schedules:
    if schedule.sa_schedule_tuple_id == power_delivery_req.sa_schedule_tuple_id:
        for sa_profile_entry in schedule.p_max_schedule.schedule_entries:
            for ev_profile_entry in power_delivery_req.charging_profile.profile_entries:
                if ev_profile_entry.start <= sa_profile_entry.time_interval.start:
                    sa_pmax = sa_profile_entry.p_max.value * 10 ** sa_profile_entry.p_max.multiplier
                    ev_pmax = ev_profile_entry.max_power.value * 10 ** ev_profile_entry.max_power.multiplier
                    if ev_pmax > sa_pmax:
                        self.stop_state_machine(...)
                        return
                else:
                    break

````

Reply:
The picture of a schedule in Marcs book, helps on visualising what you need to do:
<img width="982" alt="Bildschirmfoto 2022-10-24 um 19 40 29" src="https://user-images.githubusercontent.com/12515990/197601103-b346262d-7be1-4497-bc9a-3208e4133c8e.png">

Reply:
Done